Mathlib/Analysis/Meromorphic/NormalFormAt.lean

```diff
@@ -0,0 +1,234 @@
+      by_cases hn : 0 ≤ n
+      · left
+        rw [analyticAt_congr h₃g]
+        apply (AnalyticAt.zpow_nonneg (by fun_prop) hn).smul h₁g
+      · right
+        use hf
+        simp [this, WithTop.coe_lt_zero.2 (not_le.1 hn), h₃g.eq_of_nhds,
+          zero_zpow n (ne_of_not_le hn).symm]
+  · rintro (h | ⟨h₁, h₂, h₃⟩)
+    · by_cases h₂f : h.order = ⊤
+      · rw [AnalyticAt.order_eq_top_iff] at h₂f
+        tauto
+      · right
+        use h.order.toNat
+        have : h.order ≠ ⊤ := h₂f
+        rw [← ENat.coe_toNat_eq_self, eq_comm, AnalyticAt.order_eq_nat_iff] at this
+        obtain ⟨g, h₁g, h₂g, h₃g⟩ := this
+        use g, h₁g, h₂g
+        simpa
+    · right
+      lift h₁.order to ℤ using LT.lt.ne_top h₂ with n hn
+      obtain ⟨g, h₁g, h₂g, h₃g⟩ := (h₁.order_eq_int_iff n).1 hn.symm
+      use n, g, h₁g, h₂g
+      filter_upwards [eventually_nhdsWithin_iff.1 h₃g]
+      intro z hz
+      by_cases h₁z : z = x
+      · simp only [h₁z, h₂, Pi.smul_apply', Pi.pow_apply, sub_self]
+        rw [h₃]
+        apply (smul_eq_zero_of_left (zero_zpow n _) (g x)).symm
+        by_contra hCon
+        simp [hCon] at h₂
+      · exact hz h₁z
+
+/-- Meromorphicity in normal form is a local property. -/
+theorem meromorphicNFAt_congr {g : 𝕜 → E} (hfg : f =ᶠ[𝓝 x] g) :
+    MeromorphicNFAt f x ↔ MeromorphicNFAt g x := by
+  constructor
+  · rintro (h | ⟨n, h, h₁h, h₂h, h₃h⟩)
+    · exact Or.inl (hfg.symm.trans h)
+    · exact Or.inr ⟨n, h, h₁h, h₂h, hfg.symm.trans h₃h⟩
+  · rintro (h | ⟨n, h, h₁h, h₂h, h₃h⟩)
+    · exact Or.inl (hfg.trans h)
+    · exact Or.inr ⟨n, h, h₁h, h₂h, hfg.trans h₃h⟩
+
+/-!
+## Relation to other properties of functions
+-/
+
+/-- If a function is meromorphic in normal form at `x`, then it is meromorphic at `x`. -/
+theorem MeromorphicNFAt.meromorphicAt (hf : MeromorphicNFAt f x) :
+    MeromorphicAt f x := by
+  rw [MeromorphicAt.meromorphicNFAt_iff] at hf
+  rcases hf with h | h
+  · exact h.meromorphicAt
+  · obtain ⟨hf, _⟩ := h
+    exact hf
+
+/-- If a function is meromorphic in normal form at `x`, then it has non-negative order iff it is
+analytic. -/
+theorem MeromorphicNFAt.nonneg_order_iff_analyticAt (hf : MeromorphicNFAt f x) :
+    0 ≤ hf.meromorphicAt.order ↔ AnalyticAt 𝕜 f x := by
+  constructor <;> intro h₂f
+  · rw [MeromorphicAt.meromorphicNFAt_iff] at hf
+    rcases hf with h | ⟨_, h₃f, _⟩
+    · exact h
+    · by_contra h'
+      exact lt_irrefl 0 (lt_of_le_of_lt h₂f h₃f)
+  · rw [h₂f.meromorphicAt_order]
+    simp
+
+/-- Analytic functions are meromorphic in normal form. -/
+theorem AnalyticAt.MeromorphicNFAt (hf : AnalyticAt 𝕜 f x) :
+    MeromorphicNFAt f x := by
+  simp [MeromorphicAt.meromorphicNFAt_iff, hf]
+
+/-!
+## Continuous extension and conversion to normal form
+-/
+
+/-- Convert a meromorphic function to normal form at `x` by changing its value at `x`. -/
+noncomputable def MeromorphicAt.toNF (hf : MeromorphicAt f x) :
+    𝕜 → E := by
```

Comment:
This will display as `MeromorphicAt.toNF ⋯` in the infoview, meaning you won't know what function you are taking the normal form of. Instead I would do
```suggestion
variable (f) in
/-- Convert a meromorphic function to normal form at `x` by changing its value at `x`. -/
noncomputable def MeromorphicAt.toNF (hf : MeromorphicAt f x) :
    𝕜 → E := by
```
and by this point it's not clear why you're asking for the meromorphicity assumption anymore, so the following seems best
```suggestion
variable (f) in
/-- Convert a meromorphic function to normal form at `x` by changing its value at `x`. -/
noncomputable def toMeromorphicNFAt : 𝕜 → E :=
  if hf : MeromorphicAt f x then currentDef else 0
```

Reply:
@YaelDillies I am not 100% sure I understand. I get a lot of errors with this. Could it be that `x` must also be included someplace? I tried the following:

```
variable (f) (x) in
/-- Convert a meromorphic function to normal form at `x` by changing its value at `x`. -/
noncomputable def Function.toMeromorphicNFAt :
    𝕜 → E := by
  by_cases hf : MeromorphicAt f x
  · classical -- do not complain about decidability issues in Function.update
    apply Function.update f x
    by_cases h₁f : hf.order = (0 : ℤ)
    · rw [hf.order_eq_int_iff] at h₁f
      exact (Classical.choose h₁f) x
    · exact 0
  · exact 0
```
This also allows writing `f.toMeromorphicNFAt x`, which is maybe a little nicer as notation. What do you think?

Reply:
Yes sorry, should be mentioning `x` indeed

Reply:
Thanks for the extremely fast response. Will implement this now.

Reply:
Ok, done. Thanks for the suggestion.